org/mozilla/jss/provider/javax/crypto/JSSMacSpi.java

```diff
@@ -48,20 +54,37 @@ public void engineInit(Key key, AlgorithmParameterSpec params)
         throws InvalidKeyException, InvalidAlgorithmParameterException
     {
       try {
-        SymmetricKey real_key;
+        SymmetricKey real_key = null;
         if (key instanceof SecretKeyFacade) {
             SecretKeyFacade facade = (SecretKeyFacade)key;
             real_key = facade.key;
         } else if (key instanceof SymmetricKey) {
             real_key = (SymmetricKey)key;
-        } else {
+        } else if (key.getEncoded() != null) {
+            SecretKeyFactory factory = SecretKeyFactory.getInstance(keyName, "Mozilla-JSS");
+            SecretKeySpec spec = new SecretKeySpec(key.getEncoded(), keyName);
+            Key manufactured = factory.generateSecret(spec);
+            if (manufactured instanceof SecretKeyFacade) {
+                SecretKeyFacade facade = (SecretKeyFacade)manufactured;
+                real_key = facade.key;
+            } else if (manufactured instanceof SymmetricKey) {
+                real_key = (SymmetricKey)manufactured;
+            }
+        }
+
+        if (real_key == null) {
             throw new InvalidKeyException("Must use a key created by JSS! Try exporting the key data and importing it via SecretKeyFactory.");
         }
```

Comment:
How about something like this so we can generate more specific exception message?
```
if (key instanceof SecretKeyFacade) {
    ...
} else if (key instanceof SymmetricKey) {
    ...
} else if (key.getEncoded() != null) {
    if (manufactured instanceof SecretKeyFacade) {
        ...
    } else if (manufactured instanceof SymmetricKey) {
        ...
    } else {
        throw new InvalidKeyException("Unsupported key: " + manufactured.getClass().getName());
    }
} else {
    throw new InvalidKeyException("Missing key data: " + key.getClass().getName());
}

Reply:
Once we've imported the key to JSS, it shouldn't be unsupported any longer. It'll throw an exception (`InvalidKeySpecException`) if we can't import the key, or we get a valid key out.

Reply:
@edewata  Ok this has been updated.